### DISCUSSION TITLE
Integrate core types with interfaces

### DISCUSSION THREAD

**Discussion by sipsma** [dagger/dagger#12369] [maintainers]
Follow up to https://github.com/dagger/dagger/pull/6254

You can't yet use types from the core api like `Container`/`Directory`/etc. to implement a given interface.

You also can't implement core types through the use of interface features.
* This would be especially useful for use cases @shykes mentioned around e.g. implementing a virtual `Directory` type in a userspace module
* This should also be totally backwards compatible. If a userspace module object implements e.g. the `Directory` APIs then consumers should just be able to call `AsDirectory` on it to convert it to a `Directory` with an alternative implementation (but same type as far as SDKs are concerned)

**Comment by shykes** [comment]
I LOVE this 😁. Can't wait.
